### PR TITLE
Adding kustomize templates for k8s onboarding

### PR DESCRIPTION
**Pull request by gizas**
- Enhancement

## What does this PR do?

WHAT: Introduces the onboarding templates of kustomize for k8s observability
WHY: https://github.com/elastic/elastic-agent/issues/4657 

## Checklist

- [x] My code follows the style guidelines of this project
- [x] I have commented my code, particularly in hard-to-understand areas
- [] I have made corresponding changes to the documentation
- [x] I have made corresponding change to the default configuration files
- [ ] I have added tests that prove my fix is effective or that my feature works
- [x] I have added an entry in `./changelog/fragments` using the [changelog tool](https://github.com/elastic/elastic-agent#changelog)
- [ ] I have added an integration test or an E2E test

## How to test this PR locally

1. Clone this repo
2. Change to directory elastic/elastic-agent/deploy/kubernetes
3. Use makefile to generate templates 
```bash
GENERATEKUSTOMIZE=true make ci-create-kustomize
 ```
4. test the templates with
```bash
cd elastic-agent-kustomize/default
kubectl apply -k elastic-agent-managed
or 
kubectl apply -k elastic-agent-standalone
```

- Secrets are created here https://github.com/elastic/elastic-agent/pull/4754/files#diff-67cc02e5c01acb0ecf253b716b2c223ecd4e8366e865a3ff41b9817bfd3243f7R9
- Are referenced here https://github.com/elastic/elastic-agent/pull/4754/files#diff-9aeecc78b3187a6c2b495ae90fdf0555402dc580422a3192cd013696f742a83fR39

Examples of online references:

```bash
kubectl kustomize https://github.com/elastic/elastic-agent/deploy/kubernetes/elastic-agent-kustomize/default/elastic-agent-managed\?ref\=kustomizepoc

kubectl kustomize https://github.com/elastic/elastic-agent/deploy/kubernetes/elastic-agent-kustomize/default/elastic-agent-managed\?ref\=main

kubectl apply -k https://github.com/elastic/elastic-agent/deploy/kubernetes/elastic-agent-kustomize/default/elastic-agent-managed\?ref\=v8.13.1
```

## Testing for onboarding procedure

Managed: 
```bash
❯ kubectl kustomize elastic-agent-kustomize/default/elastic-agent-managed/ | sed -e "s/JUVOUk9MTE1FTlRfVE9LRU4l/base64_ENCODED_ENROLLMENT_TOKEN/g" -e "s/%FLEET_URL%/https:\/\/localhost:9200/g" | kubectl apply -f-
```
Standalone
```bash
❯ kubectl kustomize elastic-agent-kustomize/default/elastic-agent-standalone/ | sed -e "s/JUFQSV9LRVkl/<base64_encoded_APIKEY>/g" -e "s/%ES_HOST%/https:\/\/localhost:9200/g" -e "s/%CA_TRUSTED%/ca_trusted_fingerprint/g" -e "s/%ELASTIC_AGENT_ID%/12345/g" | kubectl apply -f-
```

Eg. 
```bash
❯ kubectl kustomize https://github.com/elastic/elastic-agent/deploy/kubernetes/elastic-agent-kustomize/default/elastic-agent-managed\?ref\=kustomizepoc |  sed -e 's/JUFQSV9LRVkl/JUFQSV9LRVkl/g' -e "s/%ES_HOST%/https:\/\/localhost:9200/g" -e "s/%CA_TRUSTED%/ca_trusted_fingerprint/g" -e "s/%ELASTIC_AGENT_ID%/12345/g" |  kubectl apply -f-
```

**Note: API_KEY  should be given as bas64 encoded values from Kibana**

Base 64 encoding decding:
```bash
❯ echo '${enrollment_token}' | base64
JHtlbnJvbGxtZW50X3Rva2VufQo=
❯ echo '${api_key}' | base64
JHthcGlfa2V5fQo=
> echo JHtlbnJvbGxtZW50X3Rva2VufQ== | base64 -D
${enrollment_token}
```

## Images

Successful enrollement in managed

Screenshot 2024-05-15 at 5 31 11 PM

Successful Installation of kube-state-metrics and 

```bash
kgp -n kube-system
NAME                                         READY   STATUS    RESTARTS       AGE
coredns-5d78c9869d-b6np9                     1/1     Running   6 (7d8h ago)   22d
coredns-5d78c9869d-p62sz                     1/1     Running   6 (7d8h ago)   22d
elastic-agent-tvskp                          1/1     Running   0              11s
etcd-kind-control-plane                      1/1     Running   0              7d8h
kindnet-qpxrb                                1/1     Running   6 (7d8h ago)   22d
kube-apiserver-kind-control-plane            1/1     Running   0              7d8h
kube-controller-manager-kind-control-plane   1/1     Running   6 (7d8h ago)   22d
kube-proxy-96hfl                             1/1     Running   6 (7d8h ago)   22d
kube-scheduler-kind-control-plane            1/1     Running   6 (7d8h ago)   22d
kube-state-metrics-5f89fb6d84-8f2sm          1/1     Running   0              11s
```

## Related issues

- Relates #https://github.com/elastic/elastic-agent/issues/4657

### CODE REVIEW

File: deploy/kubernetes/elastic-agent-kustomize/onboarding/default/elastic-agent-managed/base/elastic-agent-managed-daemonset.yaml
Comment (constanca-m) [inline deploy/kubernetes/elastic-agent-kustomize/onboarding/default/elastic-agent-managed/base/elastic-agent-managed-daemonset.yaml:6]: Wonder if all `namespace` fields should be removed from all files apart from the `kustomization.yaml`? The value gets correctly overwritten anyway, but it is not doing anything anymore.

File: deploy/kubernetes/elastic-agent-kustomize/default/elastic-agent-standalone/base/elastic-agent-standalone-daemonset.yaml
Comment (constanca-m) [inline deploy/kubernetes/elastic-agent-kustomize/default/elastic-agent-standalone/base/elastic-agent-standalone-daemonset.yaml:28]: Maybe we could have another base for hints, instead of leaving it in a comment here.

**Comment by constanca-m** [comment]
I am a bit confusing. What is the different between `default` and `onboarding` [directories](https://github.com/elastic/elastic-agent/tree/48ffca891d14e3b80f329e5ced64824f8f0ea506/deploy/kubernetes/elastic-agent-kustomize)?

File: deploy/kubernetes/elastic-agent-kustomize/default/elastic-agent-standalone/base/elastic-agent-standalone-daemonset.yaml
Comment (gizas) [reply deploy/kubernetes/elastic-agent-kustomize/default/elastic-agent-standalone/base/elastic-agent-standalone-daemonset.yaml:28]: As we discuss lets tackle this in another story

File: deploy/kubernetes/Makefile
Comment (constanca-m) [inline deploy/kubernetes/Makefile:18]: Don't forget to remove this

File: deploy/kubernetes/elastic-agent-kustomize/default/README.md
Comment (constanca-m) [inline deploy/kubernetes/elastic-agent-kustomize/default/README.md:8]: Better to link from `https://github.com/elastic/elastic-agent/...`, otherwise 404

File: deploy/kubernetes/elastic-agent-kustomize/default/README.md
Comment (constanca-m) [inline deploy/kubernetes/elastic-agent-kustomize/default/README.md:14]: Here it is better to link the file.

File: deploy/kubernetes/elastic-agent-kustomize/default/README.md
Comment (constanca-m) [inline deploy/kubernetes/elastic-agent-kustomize/default/README.md:53]: The format is not correct here, maybe it needs an empty line before the bash code?

File: deploy/kubernetes/elastic-agent-kustomize/default/README.md
Comment (constanca-m) [inline deploy/kubernetes/elastic-agent-kustomize/default/README.md:71]: Also needs another link (because this PR is not merged, maybe you need to write it as if exists in main already)

File: deploy/kubernetes/elastic-agent-kustomize/default/README.md
Comment (gizas) [reply deploy/kubernetes/elastic-agent-kustomize/default/README.md]: fixed thanks!

File: deploy/kubernetes/elastic-agent-kustomize/default/README.md
Comment (gizas) [reply deploy/kubernetes/elastic-agent-kustomize/default/README.md:14]: Done

File: deploy/kubernetes/elastic-agent-kustomize/default/README.md
Comment (constanca-m) [inline deploy/kubernetes/elastic-agent-kustomize/default/README.md:14]:
```suggestion
For *Managed Elastic Agent*, please update the following secrets inside main [kustomization.yaml](./elastic-agent-managed/kustomization.yaml):
```

File: deploy/kubernetes/elastic-agent-kustomize/default/README.md
Comment (constanca-m) [inline deploy/kubernetes/elastic-agent-kustomize/default/README.md:27]:
```suggestion
For *Standalone Elastic Agent*, please update the following secrets inside main [kustomization.yaml](./elastic-agent-standalone/kustomization.yaml):
```

File: deploy/kubernetes/elastic-agent-kustomize/default/README.md
Comment (constanca-m) [inline deploy/kubernetes/elastic-agent-kustomize/default/README.md:39]: Would this be the command our users would get from this Kibana page?

image

Comment (constanca-m) [reply deploy/kubernetes/elastic-agent-kustomize/default/README.md:39]: I think a more intuitive approach would be to change the `kustomization.yaml` file directly. So the user would see this file in Kibana:

```yaml
apiVersion: kustomize.config.k8s.io/v1beta1
kind: Kustomization

namespace: kube-system

images:
  - name: docker.elastic.co/beats/elastic-agent
    newTag: ${IMAGE_TAG}

secretGenerator:
    - name: elastic-agent-creds
      literals:
        - host=${ES_HOST} 
        - enrollment_token=${ENROLLMENT_TOKEN} 

resources:
  - ./base
  - https://github.com/kubernetes/kube-state-metrics

patches:
- path: environmental-variables-remove.yaml
- target:
    group: apps
    version: v1
    kind: DaemonSet
    name: elastic-agent
  path: fleet-enrollment-token-patch.yaml
```

And all `${...}` variables should have literal values by that point.

I think the problem with that command is that it still requires us to evaluate `base64` of the variables and we end up not making use of the literal of the secret map. And then the `sed` is not easy to read.

File: deploy/kubernetes/elastic-agent-kustomize/default/README.md
Comment (mykolaharmash) [inline deploy/kubernetes/elastic-agent-kustomize/default/README.md:35]: Not sure if this is the right name for this variable, real Elastic Agent ID is not known at the time we run `kubectl ...`, something like `DATASET_ID` might work better.

File: deploy/kubernetes/elastic-agent-kustomize/default/README.md
Comment (mykolaharmash) [inline deploy/kubernetes/elastic-agent-kustomize/default/README.md:34]: Is this needed to support self-signed certificates? We can skip it for production use-cases, right?

File: deploy/kubernetes/elastic-agent-kustomize/default/README.md
Comment (gizas) [reply deploy/kubernetes/elastic-agent-kustomize/default/README.md:43]: All of this will be replaced by onliner command and nothing will be visible to the user

File: deploy/kubernetes/elastic-agent-kustomize/default/README.md
Comment (gizas) [reply deploy/kubernetes/elastic-agent-kustomize/default/README.md:34]: Indeed it is needed only for self-signed and can be skipped for production use-cases.
We can run a test with a "production" site and prove that if we leave it that wont create any problem

File: deploy/kubernetes/elastic-agent-kustomize/default/README.md
Comment (gizas) [reply deploy/kubernetes/elastic-agent-kustomize/default/README.md:35]: To be accurate this is the session id that will be created from kibana and will denote the specific agent or in general the installation process. Yes I dont mind to change it to whatever UI team things is more appopriate

Comment (mykolaharmash) [reply deploy/kubernetes/elastic-agent-kustomize/default/README.md:35]: Awesome, thank you! `SESSION_ID` sounds to me a bit ephemeral, considering that it will be present for the whole lifespan of the agent config. `DATASET_ID` is the best I can think of now…

Oh, and btw, could we make it a string type so it supports UUID values?

Comment (gizas) [reply deploy/kubernetes/elastic-agent-kustomize/default/README.md]: Thanks done: [111d8d0](https://github.com/elastic/elastic-agent/pull/4754/commits/111d8d0a3cb47e2941faf049e8eb3c4f20051bbf)

File: deploy/kubernetes/elastic-agent-kustomize/default/README.md
Comment (mykolaharmash) [inline deploy/kubernetes/elastic-agent-kustomize/default/README.md:51]: Just to double check, should `sed -e "s/JUFQSV9LRVkl/<base64_encoded_APIKEY>/g"` be sed -e "s/%API_KEY%/<base64_encoded_APIKEY>/g"?

**Comment by gizas** [comment]
@elastic/elastic-agent-control-plane team can I have a review please?

**Comment by pierrehilbert** [comment]
@blakerouse / @andrzej-stencel could you please have a look here when you will have time?

File: deploy/kubernetes/elastic-agent-kustomize/default/README.md
Comment (mykolaharmash) [reply deploy/kubernetes/elastic-agent-kustomize/default/README.md:51]: Ah, I see, `JUFQSV9LRVkl` is indeed inside the config produced by `kustomize`. @gizas could you please clarify what is this value, can we rely on it not changing for the purpose of onboarding?

**Comment by mykolaharmash** [comment]
@gizas when I try to run this command
```
kubectl kustomize ./elastic-agent-standalone/ | sed -e "s/JUFQSV9LRVkl/V1E5ZUp.../g" -e "s/%ES_HOST%/http:\/\/localhost:9200/g" -e "s/%ELASTIC_AGENT_ID%/test-id/g" | kubectl apply -f-
```

I get this error:
```
error: error validating "STDIN": error validating data: failed to download openapi: Get "http://localhost:8080/openapi/v2?timeout=32s"
```

Is this expected? Do you have an idea what I might be missing?

File: deploy/kubernetes/elastic-agent-kustomize/default/README.md
Comment (gizas) [reply deploy/kubernetes/elastic-agent-kustomize/default/README.md:51]: Indeed you dont need to change it. 
Actually initially the kustomize will create a secrtet api_key with encoded value of JUFQSV9LRVk
The reason for this is that the creation of variables as secrtes is done in base64 
```yaml
apiVersion: v1
data:
  api_key: JUFQSV9LRVkl
kind: Secret
```

**Comment by gizas** [comment]
> I am a bit confused. What is the different between `default` and `onboarding` [directories](https://github.com/elastic/elastic-agent/tree/48ffca891d14e3b80f329e5ced64824f8f0ea506/deploy/kubernetes/elastic-agent-kustomize)?

Just for clarity, the default dir is the original dirs that we currently keep for our elastic-agent manifests
We created the onboarding folders as a copy based on he default dirs to be sure that we dont affect current folders of our users

File: deploy/kubernetes/elastic-agent-kustomize/onboarding/default/elastic-agent-managed/base/elastic-agent-managed-daemonset.yaml
Comment (gizas) [reply deploy/kubernetes/elastic-agent-kustomize/onboarding/default/elastic-agent-managed/base/elastic-agent-managed-daemonset.yaml]: I will keep it for now for clarity

File: deploy/kubernetes/elastic-agent-kustomize/default/README.md
Comment (cmacknz) [inline deploy/kubernetes/elastic-agent-kustomize/default/README.md:29]:
```suggestion
```

This doesn't look like it is supposed to be here.

File: deploy/kubernetes/elastic-agent-kustomize/default/README.md
Comment (cmacknz) [inline deploy/kubernetes/elastic-agent-kustomize/default/README.md:34]:
```suggestion
- %CA_TRUSTED%: The ssl.ca_trusted_fingerprint in order the elastic agent to be able to trust the certificate authority of the Elasticsearch output.
```

File: deploy/kubernetes/elastic-agent-kustomize/default/README.md
Comment (cmacknz) [inline deploy/kubernetes/elastic-agent-kustomize/default/README.md:33]:
```suggestion
- %API_KEY: The API Key with access privileges to connect to Elasticsearch. See [create-api-key-standalone-agent](https://www.elastic.co/guide/en/fleet/current/grant-access-to-elasticsearch.html#create-api-key-standalone-agent). *This should be encoded as base64 value because it will be stored as Kubernetes secret*
```

File: deploy/kubernetes/elastic-agent-kustomize/default/README.md
Comment (cmacknz) [inline deploy/kubernetes/elastic-agent-kustomize/default/README.md:35]: I don't understand what this does. Is this dataset in the context of the data stream naming scheme? Why is there only one?

Comment (gizas) [reply deploy/kubernetes/elastic-agent-kustomize/default/README.md:33]: In the standalone agent installations we can not verify if agents is installed and when is ready. So when a user through kibana will start a new installation (so it will have a specific session), we will include this session in the field dataset_id. And then kibana will have a loop to check whether this field has been created. 

FYI see https://github.com/elastic/elastic-agent/issues/4657 - the [Stretched goal]

File: deploy/kubernetes/elastic-agent-kustomize/default/README.md
Comment (mykolaharmash) [reply deploy/kubernetes/elastic-agent-kustomize/default/README.md:49]: Ah, I see, it's just `%API_KEY%` encoded with base64, thank you!

**Review by mykolaharmash — APPROVED** [review]
I tested the onboarding scenario and everything works as expected ✨

File: deploy/kubernetes/elastic-agent-kustomize/default/README.md
Comment (cmacknz) [reply deploy/kubernetes/elastic-agent-kustomize/default/README.md:33]: I name that doesn't repurpose the dataset terminology would make this clearer.

Also this is effectively an agent ID that isn't set by Fleet? We could just add the ability to set that so that it would populate as the `agent.id` field in documents instead of us generating a random one at startup.

Comment (gizas) [reply deploy/kubernetes/elastic-agent-kustomize/default/README.md:33]: > name that doesn't repurpose the dataset terminology would make this clearer.

@flash1293 , @mykolaharmash any other ideas regarding variable name? 

@cmacknz actually yes this is an agentId for standalone scenarios. 

> We could just add the ability to set that so that it would populate as the agent.id

When this is ready, I agree we can remove this workaround. This is I guess a quick solution now to make this working. Do you have some implementation in mind?

Comment (flash1293) [reply deploy/kubernetes/elastic-agent-kustomize/default/README.md:33]: It's not really an agent id, because it will be the same on every node of the cluster - by my understanding, the individual nodes will get assigned separate agent ids.

I agree that "dataset" is a bit confusing too, what about "onboarding id" - that's what it is after all

**Comment by tetianakravchenko** [comment]
@gizas is the description up-to-date?

> cd elastic-agent-kustomize/onboarding/default

there is no `nboarding/default` folder after executing `GENERATEKUSTOMIZE=true make ci-create-kustomize`, is it expected?

**Comment by gizas** [comment]
> there is no nboarding/default folder

Sorry typo. Updated
Folders shoudl be under `elastic-agent-kustomize/default/`

File: deploy/kubernetes/Makefile
Comment (gizas) [reply deploy/kubernetes/Makefile]: Indeed not used. I just added them because I was thinking that we should update the standalone. For now I deleted them

But it is only the meta block here https://github.com/elastic/elastic-agent/blob/main/deploy/kubernetes/elastic-agent-standalone-kubernetes.yaml#L35-L38, that actually is not used anywhere.

File: deploy/kubernetes/elastic-agent-kustomize/default/README.md
Comment (tetianakravchenko) [inline deploy/kubernetes/elastic-agent-kustomize/default/README.md:58]: here should be used `echo -n` instead, otherwise it might not work:
```
% echo -n "Zlh5ekw1QUJubU8xdUFJajNpS2M6QWdaZFIwX0xUNmFhQ3BGbUVxTVZwUQ==" | base64
WmxoNWVrdzFRVUp1YlU4eGRVRkphak5wUzJNNlFXZGFaRkl3WDB4VU5tRmhRM0JHYlVWeFRWWndVUT09

% echo "Zlh5ekw1QUJubU8xdUFJajNpS2M6QWdaZFIwX0xUNmFhQ3BGbUVxTVZwUQ==" | base64
WmxoNWVrdzFRVUp1YlU4eGRVRkphak5wUzJNNlFXZGFaRkl3WDB4VU5tRmhRM0JHYlVWeFRWWndVUT09Cg==
```

when the result of `echo <token> | base64` is used, agent fails with error:
```
Error: fail to enroll: fail to execute request to fleet-server: net/http: invalid header field value for "Authorization"
```

ideally would be to avoid this manual transformation, wondering if kustomize does not provide this option somehow

Comment (gizas) [reply deploy/kubernetes/elastic-agent-kustomize/default/README.md]: @tetianakravchenko  I update the README file for this. Indeed echo adds an extra line by default. 

During the onboarding flow the base64 values are provided from the kibana ui page

**Review by tetianakravchenko — APPROVED** [review]

**Review by constanca-m — APPROVED** [review]

File: deploy/kubernetes/elastic-agent-kustomize/default/README.md
Comment (cmacknz) [reply deploy/kubernetes/elastic-agent-kustomize/default/README.md:33]: Anything that isn't already a term used elsewhere in the product is fine with me.

Comment (gizas) [reply deploy/kubernetes/elastic-agent-kustomize/default/README.md]: Thank you both, ONBOARDING_ID added in 
[89cc0d9](https://github.com/elastic/elastic-agent/pull/4754/commits/89cc0d9062d59634111d18f6a83353f75783ef56)

**Comment by mykolaharmash** [comment]
@gizas one other thing came up about the elastic agent version. Right now in the template version is set to non-existing 8.15.0. Can we somehow ensure that we always have a valid version there, or alternatively expose is as well as variable, so for onboarding case we can set it on our side?

**Comment by gizas** [comment]
@mykolaharmash the way it works is that the [Makefile](https://github.com/elastic/elastic-agent/blob/a6dba3d77293c1cfc015b58d87203d42a09484bf/deploy/kubernetes/Makefile#L5) computes he agent image version based on this file: https://github.com/elastic/elastic-agent/blob/a6dba3d77293c1cfc015b58d87203d42a09484bf/version/version.go#L7 

So all the daemonset file has the elastic-agent:8.15.0 inside. This is only in one place

So i guess another sed like:  sed -e "s/elastic-agent:8.15.0/elastic-agent:<yourVERSION>/g" can do the magic.

Also, I think this is only now a problem until we merge the PR. After that we wont have the problem and the CI will make sure that this is always correct. And the user can choose the version with the following use of ref:
`kubectl kustomize elastic-agent-kustomize/default/elastic-agent-managed?ref=v8.15.1`

**Comment by gizas** [comment]
Updating my answer above, I think will be mandatory for kibana UI onboarding to update the link like 
`kubectl kustomize <url>?ref=v8.15.1` always with the existing version of the system of the user. Otherwise we will use always the main branch and this might break in the periods where the version.go file is ahead from the latest released versiosn

**Comment by flash1293** [comment]
@gizas Just to make sure - is there an automation that there is a version of the template for every stack version? Or is this something that needs to be done manually? We should make sure that this won't break in the next version.

**Comment by gizas** [comment]
> is there an automation that there is a version of the template for every stack version? 

Yes as part of this PR, I have added this in the makefile:
https://github.com/elastic/elastic-agent/pull/4754/files#diff-744561ff1e6cb54740e099862a03543c589ecd5416130dab47ce3c4f1b6130a1R32

**Comment by mykolaharmash** [comment]
Got it, thank you! I'll update the onboarding logic so that URL always has the version ref

**Comment by gizas** [comment]
@cmacknz or @blakerouse if no other objection can you please approve this? We will need one more to merge

**Comment by mykolaharmash** [comment]
@gizas one more things about the Elastic Agent version, am I correct that even after your new change is merged, we still need to wait for Agent v8.15.0 to be released in order to use the new template in onboarding? Because if we reference `ref=8.14.1` it will have the old one, right?

**Comment by flash1293** [comment]
Good point @mykolaharmash , ideally we could backport this so it will work for 8.14.x as well?

**Comment by gizas** [comment]
@mykolaharmash I added backport-v8.14.0 label. Thanks!
I will check the backport

@elastic/elastic-agent-control-plane team can I have an approval please?

**Comment by pierrehilbert** [comment]
@blakerouse / @andrzej-stencel could you please review here?

**Pull request merged**
